keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,203 @@
+---
+kep-number: 29
+title: Operator Dependencies
+short-desc: Introducing operators depending on other operators
+authors:
+  - "@zen-dog"
+owners:
+  - "@zen-dog"
+editor: @zen-dog
+creation-date: 2020-03-30
+last-updated: 2020-03-30
+status: provisional
+---
+
+# Operator Dependencies
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Implementation Details](#implementation-detailsnotesconstraints)
+    * [Operator Task](#operator-task)
+    * [Deployment](#deployment)
+      * [Client-Side](#client-side)
+      * [Server-Side](#server-side)
+      * [Parameterization](#dependencies-parameterization)
+    * [Update](#update)
+    * [Upgrade](#upgrade)
+    * [Uninstalling](#uninstalling)
+  * [Risks and Mitigation](#risks-and-mitigation)
+* [Alternatives](#alternatives)
+
+## Summary
+
+This KEP aims to improve operator user and developer experience by introducing operator dependencies.
+
+## Motivation
+
+Recent operator development has shown that complex operators often depend on other operators to function. One workaround commonly seen can be found in the [Flink Demo](https://github.com/kudobuilder/operators/tree/master/repository/flink/docs/demo/financial-fraud). It requires manual installation of the dependency operators while skipping the instance creation (using the `--skip-instance` option). The `Instance` is then "manually" created during the execution of the `deploy` plan. This KEP is aiming at improving this experience, streamlining the developer and user experience.
+
+### Goals
+
+Dependencies can be a complex topic. This KEP is not trying to boil the dependency ocean but rather limits itself to installation dependencies only.
+
+### Non-Goals
+
+Dependency on an already running `Instance` is a non-goal. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependency presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded and the new version is incompatible with the current Kafka `Instance`? How can compatibility be ensured? This KEP deliberately ignores this area and instead focuses on installation dependencies. Additionally, this KEP does not address output variables or referencing `Instance` resources.
```

Comment:
Maybe we should add that using the same dependency instance from two points of the dependency graph is not supported? As written below, if we have two operators that wants to install zookeeper, we need to create two zk instances and can't use a single one?

Reply:
> Maybe we should add that using the same dependency instance from two points of the dependency graph is not supported?

It is **not** as it will create a dependency circle.